lib/proxy/logging_resolv.rb

```diff
@@ -0,0 +1,63 @@
+require 'resolv'
+require 'proxy/time_utils'
+require 'forwardable'
+
+# Decorator for Resolv and DNS::Resolv Ruby classes that performs
+# logging into the smart proxy logger.
+module Proxy
+  class LoggingResolv
+    include ::Proxy::Log
+    include ::Proxy::TimeUtils
+    extend Forwardable
+
+    def_delegators :@resolv, :each_address, :each_name
+
+    def initialize(resolv)
+      @resolv = resolv
+    end
+
+    def resolver(override_nameserver = @server)
+      dns_resolv(:nameserver => override_nameserver)
+    end
+
+    def getresource(name, typeclass)
+      call_with_timing(:getresource, name, typeclass)
+    end
+
+    def getresources(name, typeclass)
+      call_with_timing(:getresources, name, typeclass)
+    end
+
+    def getname(name)
+      call_with_timing(:getname, name)
+    end
+
+    def getnames(name)
+      call_with_timing(:getnames, name)
+    end
+
+    def getaddress(name)
+      call_with_timing(:getaddress, name)
+    end
+
+    def getaddresses(name)
+      call_with_timing(:getaddresses, name)
+    end
+
+    private
+
+    def call_with_timing(method, *args)
+      result = nil
+      duration = time_spent_in_ms do
+        result = @resolv.send(method, *args)
+      end
+      if duration < 9000
```

Comment:
Small nit: I like constants for this. Perhaps a `SLOW_DNS_QUERY_MS` definition? Note I'm not asking for a config value, just a constant in the class.

Reply:
Sure, it's probably an overkill but with all the issues, cases and confused users around DNS queries I want to be as clear as possible when this happens.